pkg/apis/autoscaling/validation/vpa.go

```diff
@@ -0,0 +1,188 @@
+	allErrs := field.ErrorList{}
+	if resourceName != nil && !supportedResources.Has(string(*resourceName)) {
+		allErrs = append(allErrs, field.NotSupported(fldPath, resourceName, supportedResources.List()))
+	}
+	return allErrs
+}
+
+func validatePodUpdatePolicy(podUpdatePolicy *autoscaling.PodUpdatePolicy, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+	if podUpdatePolicy != nil {
+		allErrs = append(allErrs, validateUpdateMode(&podUpdatePolicy.UpdateMode, fldPath.Child("updateMode"))...)
+	}
+	return allErrs
+}
+
+// Verifies that the core.ResourceList contains valid and supported resources (see supportedResources).
+// If upperBound != nil, additionally checks that the quantity of resources in resourceList does not exceed
+// the corresponding quantity in upperBound.
+func validateResourceList(resourceList *core.ResourceList, upperBound *core.ResourceList, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+	if resourceList != nil {
+		for resourceName, quantity := range *resourceList {
+			resPath := fldPath.Key(string(resourceName))
+			// Validate resource name.
+			allErrs = append(allErrs, validateResourceName(&resourceName, resPath)...)
+			// Validate resource quantity.
+			allErrs = append(allErrs, corevalidation.ValidateResourceQuantityValue(string(resourceName), quantity, resPath)...)
+			if upperBound != nil {
+				// Check that request <= limit.
+				upperBoundQuantity, exists := (*upperBound)[resourceName]
+				if exists && quantity.Cmp(upperBoundQuantity) > 0 {
+					allErrs = append(allErrs, field.Invalid(fldPath, quantity.String(),
+						"must be less than or equal to the upper bound"))
+				}
+			}
+		}
+	}
+	return allErrs
+}
+
+func validateContainerResourcePolicy(containerResourcePolicy *autoscaling.ContainerResourcePolicy, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+	if containerResourcePolicy != nil {
+		allErrs = append(allErrs, validateContainerScalingMode(&containerResourcePolicy.Mode, fldPath.Child("mode"))...)
+		allErrs = append(allErrs, validateResourceList(&containerResourcePolicy.MinAllowed, &containerResourcePolicy.MaxAllowed, fldPath.Child("minAllowed"))...)
+		allErrs = append(allErrs, validateResourceList(&containerResourcePolicy.MaxAllowed, nil, fldPath.Child("maxAllowed"))...)
+	}
+	return allErrs
+}
+
+func validatePodResourcePolicy(podResourcePolicy *autoscaling.PodResourcePolicy, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+	if podResourcePolicy != nil {
+		for i, containerPolicy := range podResourcePolicy.ContainerPolicies {
+			allErrs = append(allErrs, validateContainerResourcePolicy(&containerPolicy, fldPath.Child("containerPolicies").Index(i))...)
+		}
+	}
+	return allErrs
+}
+
+func validateVerticalPodAutoscalerSpec(spec *autoscaling.VerticalPodAutoscalerSpec, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+	if spec != nil {
+		if spec.Selector != nil {
```

Comment:
usually I see this flipped.  Required test then valid test.  Minor.

Reply:
Done.